apollo-federation/src/operation/mod.rs

```diff
@@ -3042,8 +3042,17 @@ impl SelectionSet {
         })
     }
 
+    /// In a normalized selection set containing only fields and inline fragments,
+    /// iterate over all the fields that may be selected.
+    ///
+    /// # Preconditions
+    /// The selection set must not contain named fragment spreads.
+    pub(crate) fn field_selections(&self) -> FieldSelectionsIter<'_> {
+        FieldSelectionsIter::new(self.selections.values())
+    }
```

Comment:
We do not use the whole `CollectedFieldInSet` data structure in `optimize()` so we can use an iterator (that has a small temporary vec inside) instead of a big temporary vec.